scipy/signal/_upfirdn.py

```diff
@@ -0,0 +1,177 @@
+# EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+# PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+# LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+# NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+# SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import numpy as np
+
+from ._upfirdn_apply import _output_len, _apply
+
+__all__ = ['upfirdn', '_output_len']
+
+
+def _pad_h(h, up):
+    """Store coefficients in a transposed, flipped arrangement.
+
+    For example, suppose upRate is 3, and the
+    input number of coefficients is 10, represented as h[0], ..., h[9].
+
+    Then the internal buffer will look like this::
+
+       h[9], h[6], h[3], h[0],   // flipped phase 0 coefs
+       0,    h[7], h[4], h[1],   // flipped phase 1 coefs (zero-padded)
+       0,    h[8], h[5], h[2],   // flipped phase 2 coefs (zero-padded)
+
+    """
+    h_padlen = len(h) + (-len(h) % up)
+    h_full = np.zeros(h_padlen, h.dtype)
+    h_full[:len(h)] = h
+    h_full = h_full.reshape(-1, up).T[:, ::-1].ravel()
+    return h_full
+
+
+class UpFIRDown(object):
+    def __init__(self, h, x_dtype, up, down):
+        """Helper for resampling"""
+        h = np.asarray(h)
+        if h.ndim != 1 or h.size == 0:
+            raise ValueError('h must be 1D with non-zero length')
+        self._output_type = np.result_type(h.dtype, x_dtype, np.float32)
+        h = np.asarray(h, self._output_type)
+        self._up = int(up)
+        self._down = int(down)
+        if self._up < 1 or self._down < 1:
+            raise ValueError('Both up and down must be >= 1')
+        # This both transposes, and "flips" each phase for filtering
+        self._h_trans_flip = _pad_h(h, self._up)
+
+    def apply_filter(self, x):
+        """Apply the prepared filter to a 1D signal x"""
+        output_len = _output_len(len(self._h_trans_flip), len(x),
+                                 self._up, self._down)
+        out = np.zeros(output_len, dtype=self._output_type)
+        _apply(np.asarray(x, self._output_type), self._h_trans_flip, out,
+               self._up, self._down)
+        return out
+
+
+def upfirdn(h, x, up=1, down=1, axis=-1):
+    """Upsample, FIR filter, and downsample
+
+    Parameters
+    ----------
+    h : array-like
+        1-dimensional FIR (finite-impulse response) filter coefficients.
+    x : array-like
+        Input signal array.
+    up : int, optional
+        Upsampling rate. Default is 1.
+    down : int, optional
+        Downsampling rate. Default is 1.
+    axis : int, optional
+        The axis of the input data array along which to apply the
+        linear filter. The filter is applied to each subarray along
+        this axis. Default is -1.
+
+    Returns
+    -------
+    y : ndarray
+        The output signal array. Dimensions will be the same as ``x`` except
```

Comment:
I've seen that but got in the habit of doing double-backticks in other packages since it seemed like the variable names never rendered properly (unless they're meant to render just in italics?) in other packages. I'll change them here though.